File: node/packages/fluent-cards/tests/advanced.test.ts
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import {
  AdaptiveCardBuilder,
  AuthenticationBuilder,
  RefreshBuilder,
} from 'fluent-cards';
import type {
  ExecuteAction,
  TokenExchangeResource,
  AuthCardButton,
} from 'fluent-cards';

describe('Version and schema URL handling', () => {
  it('withVersion("1.0") maps to 1.0.0 schema', () => {
    const card = AdaptiveCardBuilder.create().withVersion('1.0').build();
    assert.equal(card.version, '1.0');
    assert.equal(card['$schema'], 'https://adaptivecards.io/schemas/1.0.0/adaptive-card.json');
  });

  it('withVersion("1.1") maps to 1.1.0 schema', () => {
    const card = AdaptiveCardBuilder.create().withVersion('1.1').build();
    assert.equal(card.version, '1.1');
    assert.equal(card['$schema'], 'https://adaptivecards.io/schemas/1.1.0/adaptive-card.json');
  });

  it('withVersion("1.2") maps to 1.2.0 schema', () => {
    const card = AdaptiveCardBuilder.create().withVersion('1.2').build();
    assert.equal(card.version, '1.2');
    assert.equal(card['$schema'], 'https://adaptivecards.io/schemas/1.2.0/adaptive-card.json');
  });

  it('withVersion("1.3") maps to 1.3.0 schema', () => {
    const card = AdaptiveCardBuilder.create().withVersion('1.3').build();
    assert.equal(card.version, '1.3');
    assert.equal(card['$schema'], 'https://adaptivecards.io/schemas/1.3.0/adaptive-card.json');
  });

  it('withVersion("1.4") maps to 1.4.0 schema', () => {
    const card = AdaptiveCardBuilder.create().withVersion('1.4').build();
    assert.equal(card.version, '1.4');
    assert.equal(card['$schema'], 'https://adaptivecards.io/schemas/1.4.0/adaptive-card.json');
  });

  it('withVersion("1.5") maps to 1.5.0 schema', () => {
    const card = AdaptiveCardBuilder.create().withVersion('1.5').build();
    assert.equal(card.version, '1.5');
    assert.equal(card['$schema'], 'https://adaptivecards.io/schemas/1.5.0/adaptive-card.json');
  });

  it('withVersion("1.6") maps to 1.6.0 schema', () => {
    const card = AdaptiveCardBuilder.create().withVersion('1.6').build();
    assert.equal(card.version, '1.6');
    assert.equal(card['$schema'], 'https://adaptivecards.io/schemas/1.6.0/adaptive-card.json');
  });

  it('unknown version falls back to generic schema URL', () => {
    const card = AdaptiveCardBuilder.create().withVersion('9.9').build();
    assert.equal(card.version, '9.9');
    assert.equal(card['$schema'], 'http://adaptivecards.io/schemas/adaptive-card.json');
  });

  it('default version is 1.5', () => {
    const card = AdaptiveCardBuilder.create().build();
    assert.equal(card.version, '1.5');
    assert.equal(card['$schema'], 'https://adaptivecards.io/schemas/1.5.0/adaptive-card.json');
  });

  it('withSchema overrides auto-mapped URL', () => {
    const card = AdaptiveCardBuilder.create()
      .withVersion('1.6')
      .withSchema('https://custom.schema/v1')
      .build();
    assert.equal(card.version, '1.6');
    assert.equal(card['$schema'], 'https://custom.schema/v1');
  });

  it('withSchema(undefined) removes schema', () => {
    const card = AdaptiveCardBuilder.create()
      .withVersion('1.6')
      .withSchema(undefined)
      .build();
    assert.equal(card['$schema'], undefined);
  });
});

describe('Authentication builder', () => {
  it('builds with text and connectionName', () => {
    const auth = new AuthenticationBuilder()
      .withText('Sign in to continue')
      .withConnectionName('myConnection')
      .build();

    assert.equal(auth.text, 'Sign in to continue');
    assert.equal(auth.connectionName, 'myConnection');
  });

  it('builds with tokenExchangeResource', () => {
    const resource: TokenExchangeResource = {
      id: 'res1',
      uri: 'api://botid/token',
      providerId: 'azure',
    };
    const auth = new AuthenticationBuilder()
      .withTokenExchangeResource(resource)
      .build();

    assert.deepEqual(auth.tokenExchangeResource, resource);
  });

  it('builds with buttons', () => {
    const button1: AuthCardButton = {
      type: 'signin',
      title: 'Sign In',
      image: 'https://example.com/icon.png',
      value: 'https://example.com/auth',
    };
    const button2: AuthCardButton = {
      type: 'openUrl',
      title: 'Register',
      value: 'https://example.com/register',
    };
    const auth = new AuthenticationBuilder()
      .addButton(button1)
      .addButton(button2)
      .build();

    assert.equal(auth.buttons!.length, 2);
    assert.deepEqual(auth.buttons![0], button1);
    assert.deepEqual(auth.buttons![1], button2);
  });

  it('builds with all properties', () => {
    const auth = new AuthenticationBuilder()
      .withText('Auth text')
      .withConnectionName('conn')
      .withTokenExchangeResource({ id: 'r1', uri: 'uri', providerId: 'p' })
      .addButton({ type: 'signin', title: 'Sign In', value: 'url' })
      .build();

    assert.equal(auth.text, 'Auth text');
    assert.equal(auth.connectionName, 'conn');
    assert.ok(auth.tokenExchangeResource);
    assert.equal(auth.buttons!.length, 1);
  });

  it('via card builder withAuthentication', () => {
    const card = AdaptiveCardBuilder.create()
      .withAuthentication((a) =>
        a
          .withText('Please sign in')
          .withConnectionName('OAuth2')
          .addButton({ type: 'signin', title: 'SSO', value: 'https://sso.example.com' }),
      )
      .build();

    assert.ok(card.authentication);
    assert.equal(card.authentication!.text, 'Please sign in');
    assert.equal(card.authentication!.connectionName, 'OAuth2');
    assert.equal(card.authentication!.buttons!.length, 1);
  });
});

describe('Refresh builder', () => {
  it('builds with action and userIds', () => {
    const refresh = new RefreshBuilder()
      .withAction((a) => a.execute('Refresh').withVerb('refresh'))
      .addUserId('user1')
      .addUserId('user2')
      .build();

    assert.ok(refresh.action);
    assert.equal(refresh.action!.type, 'Action.Execute');
    assert.equal((refresh.action as ExecuteAction).verb, 'refresh');
    assert.deepEqual(refresh.userIds, ['user1', 'user2']);
  });

  it('builds with expires', () => {
    const refresh = new RefreshBuilder()
      .withAction((a) => a.execute('Refresh'))
      .withExpires('2024-12-31T23:59:59Z')
      .build();

    assert.equal(refresh.expires, '2024-12-31T23:59:59Z');
  });

  it('via card builder withRefresh', () => {
    const card = AdaptiveCardBuilder.create()
      .withRefresh((r) =>
        r
          .withAction((a) => a.execute('Auto Refresh').withVerb('autoRefresh'))
          .addUserId('user-abc')
          .withExpires('2025-01-01T00:00:00Z'),
      )
      .build();

    assert.ok(card.refresh);
    assert.ok(card.refresh!.action);
    assert.equal(card.refresh!.action!.type, 'Action.Execute');
    assert.deepEqual(card.refresh!.userIds, ['user-abc']);
    assert.equal(card.refresh!.expires, '2025-01-01T00:00:00Z');
  });
});

describe('Metadata', () => {
  it('sets webUrl via withMetadata', () => {
    const card = AdaptiveCardBuilder.create()
      .withMetadata('https://example.com/deep-link')
      .build();

    assert.deepEqual(card.metadata, { webUrl: 'https://example.com/deep-link' });
  });

  it('metadata is undefined when not set', () => {
    const card = AdaptiveCardBuilder.create().build();
    assert.equal(card.metadata, undefined);
  });
});

File: go/fluentcards/advanced_test.go
package fluentcards

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// ── Version handling ─────────────────────────────────────────────────────────

func TestAdvanced_Version_AllKnown(t *testing.T) {
	t.Parallel()
	versions := []struct {
		ver    string
		schema string
	}{
		{"1.0", "https://adaptivecards.io/schemas/1.0.0/adaptive-card.json"},
		{"1.1", "https://adaptivecards.io/schemas/1.1.0/adaptive-card.json"},
		{"1.2", "https://adaptivecards.io/schemas/1.2.0/adaptive-card.json"},
		{"1.3", "https://adaptivecards.io/schemas/1.3.0/adaptive-card.json"},
		{"1.4", "https://adaptivecards.io/schemas/1.4.0/adaptive-card.json"},
		{"1.5", "https://adaptivecards.io/schemas/1.5.0/adaptive-card.json"},
		{"1.6", "https://adaptivecards.io/schemas/1.6.0/adaptive-card.json"},
	}
	for _, tc := range versions {
		card := NewAdaptiveCardBuilder().WithVersion(tc.ver).Build()
		assert.Equal(t, tc.ver, card["version"], "version %s", tc.ver)
		assert.Equal(t, tc.schema, card["$schema"], "schema for version %s", tc.ver)
	}
}

func TestAdvanced_Version_Unknown(t *testing.T) {
	t.Parallel()
	card := NewAdaptiveCardBuilder().WithVersion("9.9").Build()
	assert.Equal(t, "9.9", card["version"])
	assert.Equal(t, "http://adaptivecards.io/schemas/adaptive-card.json", card["$schema"])
}

func TestAdvanced_Version_SchemaOverride(t *testing.T) {
	t.Parallel()
	card := NewAdaptiveCardBuilder().
		WithVersion("1.5").
		WithSchema("https://custom.schema/v1").
		Build()
	assert.Equal(t, "https://custom.schema/v1", card["$schema"])
}

// ── Authentication ──────────────────────────────────────────────────────────

func TestAdvanced_Authentication_Full(t *testing.T) {
	t.Parallel()
	card := NewAdaptiveCardBuilder().
		WithAuthentication(func(auth *AuthenticationBuilder) {
			auth.
				WithText("Please sign in").
				WithConnectionName("myConnection").
				WithTokenExchangeResource(map[string]any{
					"id":  "res1",
					"uri": "https://token.example.com",
				}).
				AddButton(map[string]any{
					"type":  "signin",
					"title": "Sign In",
					"value": "https://login.example.com",
				})
		}).
		Build()

	authData := card["authentication"].(Card)
	assert.Equal(t, "Please sign in", authData["text"])
	assert.Equal(t, "myConnection", authData["connectionName"])

	ter := authData["tokenExchangeResource"].(map[string]any)
	assert.Equal(t, "res1", ter["id"])
	assert.Equal(t, "https://token.example.com", ter["uri"])

	buttons := authData["buttons"].([]any)
	require.Len(t, buttons, 1)
	btn := buttons[0].(map[string]any)
	assert.Equal(t, "signin", btn["type"])
	assert.Equal(t, "Sign In", btn["title"])
}

func TestAdvanced_Authentication_MultipleButtons(t *testing.T) {
	t.Parallel()
	card := NewAdaptiveCardBuilder().
		WithAuthentication(func(auth *AuthenticationBuilder) {
			auth.
				AddButton(map[string]any{"type": "signin", "title": "SSO"}).
				AddButton(map[string]any{"type": "signin", "title": "OAuth"})
		}).
		Build()

	authData := card["authentication"].(Card)
	buttons := authData["buttons"].([]any)
	require.Len(t, buttons, 2)
}

// ── Refresh ─────────────────────────────────────────────────────────────────

func TestAdvanced_Refresh_Full(t *testing.T) {
	t.Parallel()
	card := NewAdaptiveCardBuilder().
		WithRefresh(func(r *RefreshBuilder) {
			r.
				WithAction(func(a *ActionBuilder) {
					a.Execute("Refresh").WithVerb("refresh")
				}).
				AddUserID("user1").
				AddUserID("user2").
				WithExpires("2024-12-31T23:59:59Z")
		}).
		Build()

	refresh := card["refresh"].(Card)
	action := refresh["action"].(Card)
	assert.Equal(t, "Action.Execute", action["type"])
	assert.Equal(t, "refresh", action["verb"])

	userIds := refresh["userIds"].([]any)
	require.Len(t, userIds, 2)
	assert.Equal(t, "user1", userIds[0])
	assert.Equal(t, "user2", userIds[1])

	assert.Equal(t, "2024-12-31T23:59:59Z", refresh["expires"])
}

func TestAdvanced_Refresh_ActionOnly(t *testing.T) {
	t.Parallel()
	card := NewAdaptiveCardBuilder().
		WithRefresh(func(r *RefreshBuilder) {
			r.WithAction(func(a *ActionBuilder) {
				a.Execute("Refresh")
			})
		}).
		Build()

	refresh := card["refresh"].(Card)
	assert.NotNil(t, refresh["action"])
	assert.Nil(t, refresh["userIds"])
	assert.Nil(t, refresh["expires"])
}

// ── Metadata ────────────────────────────────────────────────────────────────

func TestAdvanced_Metadata(t *testing.T) {
	t.Parallel()
	card := NewAdaptiveCardBuilder().
		WithMetadata("https://example.com/card-meta").
		Build()

	meta := card["metadata"].(Card)
	assert.Equal(t, "https://example.com/card-meta", meta["webUrl"])
}

// ── FallbackText, Speak, Lang, RTL, MinHeight ───────────────────────────────

func TestAdvanced_FallbackText(t *testing.T) {
	t.Parallel()
	card := NewAdaptiveCardBuilder().WithFallbackText("Cannot display card").Build()
	assert.Equal(t, "Cannot display card", card["fallbackText"])
}

func TestAdvanced_Speak(t *testing.T) {
	t.Parallel()
	card := NewAdaptiveCardBuilder().WithSpeak("Hello from the card").Build()
	assert.Equal(t, "Hello from the card", card["speak"])
}

func TestAdvanced_Lang(t *testing.T) {
	t.Parallel()
	card := NewAdaptiveCardBuilder().WithLang("fr").Build()
	assert.Equal(t, "fr", card["lang"])
}

func TestAdvanced_RTL(t *testing.T) {
	t.Parallel()
	card := NewAdaptiveCardBuilder().WithRTL(true).Build()
	assert.Equal(t, true, card["rtl"])
}

func TestAdvanced_RTL_False(t *testing.T) {
	t.Parallel()
	card := NewAdaptiveCardBuilder().WithRTL(false).Build()
	assert.Equal(t, false, card["rtl"])
}

func TestAdvanced_MinHeight(t *testing.T) {
	t.Parallel()
	card := NewAdaptiveCardBuilder().WithMinHeight("500px").Build()
	assert.Equal(t, "500px", card["minHeight"])
}

// ── SelectAction ────────────────────────────────────────────────────────────

func TestAdvanced_CardSelectAction(t *testing.T) {
	t.Parallel()
	card := NewAdaptiveCardBuilder().
		WithSelectAction(func(a *ActionBuilder) {
			a.OpenURL("https://example.com").WithTitle("Open Site")
		}).
		Build()

	sa := card["selectAction"].(Card)
	assert.Equal(t, "Action.OpenUrl", sa["type"])
	assert.Equal(t, "https://example.com", sa["url"])
	assert.Equal(t, "Open Site", sa["title"])
}

func TestAdvanced_CardSelectAction_Execute(t *testing.T) {
	t.Parallel()
	card := NewAdaptiveCardBuilder().
		WithSelectAction(func(a *ActionBuilder) {
			a.Execute("Tap").WithVerb("onTap").WithData(map[string]any{"x": 1})
		}).
		Build()

	sa := card["selectAction"].(Card)
	assert.Equal(t, "Action.Execute", sa["type"])
	assert.Equal(t, "onTap", sa["verb"])
	data := sa["data"].(map[string]any)
	assert.Equal(t, 1, data["x"])
}

// ── VerticalContentAlignment ────────────────────────────────────────────────

func TestAdvanced_VerticalContentAlignment(t *testing.T) {
	t.Parallel()
	for _, align := range []VerticalAlignment{VerticalAlignmentTop, VerticalAlignmentCenter, VerticalAlignmentBottom} {
		card := NewAdaptiveCardBuilder().WithVerticalContentAlignment(align).Build()
		assert.Equal(t, string(align), card["verticalContentAlignment"])
	}
}

// ── BackgroundImage ─────────────────────────────────────────────────────────

func TestAdvanced_BackgroundImage_Full(t *testing.T) {
	t.Parallel()
	card := NewAdaptiveCardBuilder().
		WithBackgroundImage(func(bi *BackgroundImageBuilder) {
			bi.
				WithURL("https://example.com/bg.png").
				WithFillMode(BackgroundImageFillModeRepeat).
				WithHorizontalAlignment(HorizontalAlignmentCenter).
				WithVerticalAlignment(VerticalAlignmentBottom)
		}).
		Build()

	bg := card["backgroundImage"].(Card)
	assert.Equal(t, "https://example.com/bg.png", bg["url"])
	assert.Equal(t, string(BackgroundImageFillModeRepeat), bg["fillMode"])
	assert.Equal(t, string(HorizontalAlignmentCenter), bg["horizontalAlignment"])
	assert.Equal(t, string(VerticalAlignmentBottom), bg["verticalAlignment"])
}

// ── Combined advanced card ──────────────────────────────────────────────────

func TestAdvanced_FullCard(t *testing.T) {
	t.Parallel()
	card := NewAdaptiveCardBuilder().
		WithVersion("1.6").
		WithFallbackText("fallback").
		WithSpeak("speak").
		WithLang("en").
		WithRTL(false).
		WithMinHeight("200px").
		WithMetadata("https://example.com").
		WithAuthentication(func(auth *AuthenticationBuilder) {
			auth.WithText("Sign in").WithConnectionName("conn")
		}).
		WithRefresh(func(r *RefreshBuilder) {
			r.WithAction(func(a *ActionBuilder) {
				a.Execute("Refresh")
			}).AddUserID("u1")
		}).
		WithSelectAction(func(a *ActionBuilder) {
			a.OpenURL("https://example.com")
		}).
		AddTextBlock(func(tb *TextBlockBuilder) {
			tb.WithText("Title").WithSize(TextSizeExtraLarge)
		}).
		AddAction(func(a *ActionBuilder) {
			a.Submit("OK")
		}).
		Build()

	assert.Equal(t, "1.6", card["version"])
	assert.Equal(t, "fallback", card["fallbackText"])
	assert.Equal(t, "speak", card["speak"])
	assert.Equal(t, "en", card["lang"])
	assert.Equal(t, false, card["rtl"])
	assert.Equal(t, "200px", card["minHeight"])
	assert.NotNil(t, card["metadata"])
	assert.NotNil(t, card["authentication"])
	assert.NotNil(t, card["refresh"])
	assert.NotNil(t, card["selectAction"])

	body := card["body"].([]any)
	require.Len(t, body, 1)

	actions := card["actions"].([]any)
	require.Len(t, actions, 1)
}
